### DISCUSSION TITLE
How to implement proper explicit feedback calculation?

### DISCUSSION THREAD

**Discussion by LennartKeuters** [hathach/tinyusb#2327] [Q&A] [answered]
Hi everybody,

First, I would like to say thank you for this great piece of work! 

I’m currently trying to implement an USB UAC2 soundcard with a STM32F446 µC and an I²S Audio-Codec on a custom board. Following the given examples, I’ve implemented the basic playback functionality using DMA to output the data via I²S.
Because of my given hardware with a dedicated master clock, I need to implement an asynchronous sink and use explicit feedback to sync the audio data with the host. So far, I have experimented a bit with tud_audio_fb_set() to set the feedback value as user of the TinyUSB stack.

Having read several issues and discussions here about explicit feedback, I’m a bit confused and don’t really know what the best solution is to do it properly. Count the frames with respect to the SOF interval, use the FIFO threshold as indicator or maybe a mixture of both?
I have seen in the source code of audio_device.c there are several functions regarding this. Unfortunately, I don’t fully understand yet which to use when and how. What functionality is currently implemented in the TinyUSB stack and what needs to be implemented by me as the user to make it work?

It would be great to have a little explanation or maybe a example covering this topic.

@HiFiPhile It seems you have implemented a working project with your i2s_bridge using the FIFO threshold? Would it be possible to get access to it?

Many thanks in advance!

**Comment by HiFiPhile** [comment, marked as answer]
You are asking at the right timing !

I just wrapped up my work in https://github.com/hathach/tinyusb/pull/2328

**Reply by LennartKeuters** [reply]
Thank you so much!
Works like a charm.
